x/consensus/operations.go

```diff
@@ -0,0 +1,107 @@
+package consensus
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/forbole/bdjuno/database"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/desmos-labs/juno/config"
+	"github.com/desmos-labs/juno/db"
+	"github.com/desmos-labs/juno/parse/client"
+	constypes "github.com/forbole/bdjuno/x/consensus/types"
+	"github.com/rs/zerolog/log"
+	tmctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// ListenOperation allows to start listening to new consensus events properly
+func ListenOperation(_ config.Config, _ *codec.Codec, cp client.ClientProxy, db db.Database) error {
+	bdDatabase, ok := db.(database.BigDipperDb)
+	if !ok {
+		return fmt.Errorf("given database instance is not a BigDipperDb")
+	}
+
+	events := []string{
+		tmtypes.EventCompleteProposal,
+		tmtypes.EventNewRound,
+		tmtypes.EventNewRoundStep,
+		tmtypes.EventPolka,
+		tmtypes.EventValidBlock,
+		tmtypes.EventVote,
+	}
+
+	var channels []<-chan tmctypes.ResultEvent
+	for index, event := range events {
+		channels = append(channels, subscribeConsensusEvent(index, event, cp))
+	}
+
+	merged := merge(channels...)
+	go func() {
+		for event := range merged {
+
+			// Serialize the event data as JSON to later de-serialize it into our custom object
+			bz, err := json.Marshal(event.Data)
+			if err != nil {
+				log.Fatal().Err(err).Send()
+			}
+
+			// De-serialize the data into our custom object
+			var consEvent constypes.ConsensusEvent
+			err = json.Unmarshal(bz, &consEvent)
+			if err != nil {
+				log.Fatal().Err(err).Send()
+			}
+
+			// Save the event
+			err = bdDatabase.SaveConsensus(consEvent)
+			if err != nil {
+				log.Fatal().Err(err).Send()
+			}
+		}
+	}()
+
+	return nil
+}
+
+// subscribeConsensusEvent allows to subscribe to the consensus event having the given name,
+// and returns a read-only channel emitting all the events
+func subscribeConsensusEvent(index int, event string, cp client.ClientProxy) <-chan tmctypes.ResultEvent {
+	query := fmt.Sprintf("tm.event = '%s'", event)
+	subscriber := fmt.Sprintf("juno-client-consensus-%d", index)
+
+	eventCh, cancel, err := cp.SubscribeEvents(subscriber, query)
+	defer cancel()
+
+	if err != nil {
+		log.Fatal().Err(err).Send()
+	}
+
+	return eventCh
+}
+
+// merge takes a list of read-only channels and merges them into a single read-only channel
+func merge(cs ...<-chan tmctypes.ResultEvent) <-chan tmctypes.ResultEvent {
+	out := make(chan tmctypes.ResultEvent)
+
+	var wg sync.WaitGroup
+	wg.Add(len(cs))
+
+	for _, c := range cs {
+		go func(c <-chan tmctypes.ResultEvent) {
```

Comment:
Why not putting subscribeConsensusEvent here? That will be more efficient

Reply:
I think that what you suggest cannot be done. This line is inside the following method: 

```go
// merge takes a list of read-only channels and merges them into a single read-only channel
func merge(cs ...<-chan tmctypes.ResultEvent) <-chan tmctypes.ResultEvent
```

There we are iterating over the list of channels and merging all their emissions into one single channel. By doing so we can than handle easily all the emitted events as the same way as they had been emitted by a single source instead of multiple subscriptions. 

This is done due to the fact that in order to listen to the current consensus, we need to listen to multiple events using the RPC, but for each event we want to listen to, we need to create a subscription. Each subscription returns a read-only channel (a stream of events). When you have multiple channels, the only way to listen to them at the same time and get all the events as soon as they are emitted preserving their order, is to merge such channels into one (at least that's what I found out). 

If you have an example implementation that works as you described I would be more than happy to see it as I didn't find anything different from this